core/jvm/src/main/scala/fs2/interop/flow/StreamSubscription.scala

```diff
@@ -96,75 +107,75 @@ private[flow] final class StreamSubscription[F[_], A] private (
-  private def cancelMe: F[Unit] =
-    canceled.complete(()).void
-
-  override def cancel(): Unit =
-    try
-      requestDispatcher.unsafeRunAndForget(cancelMe)
-    catch {
-      case _: IllegalStateException =>
-      // Dispatcher already shutdown, we are on terminal state, NOOP.
+  override def cancel(): Unit = {
+    var cancelMe = cancelToken.get()
+
+    if (cancelMe ne null) {
+      // Loop until we get the actual cancel callback.
+      while (cancelMe eq StreamSubscription.defaultCallback)
+        cancelMe = cancelToken.get()
+
+      cancelToken.set(null)
+      cancelMe()
     }
+  }
 
-  override def request(n: Long): Unit = {
-    val prog =
-      canceled.tryGet.flatMap {
-        case None =>
-          if (n == java.lang.Long.MAX_VALUE)
-            requests.offer(StreamSubscription.Request.Infinite)
-          else if (n > 0)
-            requests.offer(StreamSubscription.Request.Finite(n))
-          else
-            onError(
-              ex = new IllegalArgumentException(s"Invalid number of elements [${n}]")
-            )
-
-        case Some(()) =>
-          F.unit
+  override def request(n: Long): Unit =
+    if (cancelToken.get() ne null) {
+      if (n <= 0)
+        onError(
+          ex = new IllegalArgumentException(s"Invalid number of elements [${n}]")
+        )
+      else {
+        requests.updateAndGet { r =>
+          val result = r + n
+          if (result < 0) {
+            // Overflow
+            Long.MaxValue
+          } else {
+            result
+          }
+        }
+
+        resume.get().apply()
       }
-    try
-      requestDispatcher.unsafeRunAndForget(prog)
-    catch {
-      case _: IllegalStateException =>
-      // Dispatcher already shutdown, we are on terminal state, NOOP.
     }
-  }
 }
 
 private[flow] object StreamSubscription {
-
-  /** Represents a downstream subscriber's request to publish elements. */
-  private sealed trait Request
-  private object Request {
-    case object Infinite extends Request
-    final case class Finite(n: Long) extends Request
-  }
+  private val defaultCallback = () => ()
 
   // Mostly for testing purposes.
   def apply[F[_], A](stream: Stream[F, A], subscriber: Subscriber[A])(implicit
       F: Async[F]
-  ): Resource[F, StreamSubscription[F, A]] =
+  ): F[StreamSubscription[F, A]] =
     (
-      Dispatcher.sequential[F](await = true),
-      Resource.eval(Queue.unbounded[F, Request]),
-      Resource.eval(Deferred[F, Unit])
-    ).mapN { case (requestDispatcher, requests, canceled) =>
+      F.delay(new AtomicLong(0L)),
+      F.delay(new AtomicReference(defaultCallback)),
+      F.delay(new AtomicReference(defaultCallback)).map { cancelToken =>
+        val canceled = F.async_[Unit] { cb =>
+          cancelToken.set(() => cb(Either.unit))
+        }
+
+        cancelToken -> canceled
+      }
+    ).mapN { case (requests, resume, (cancelToken, canceled)) =>
       new StreamSubscription(
         stream,
         subscriber,
```

Comment:
Is there much value in adding a few `F.delay`? The only difference of this overhead is to allow the dispatcher to shift to another task, but would we want that? 

```suggestion
      F.delay { 
        val requests = new AtomicLong(0L)
        val resume = new AtomicReference(defaultCallback)
        val cancelToken = new AtomicReference(defaultCallback)
        val canceled = F.async_[Unit] { cb =>
          cancelToken.set(() => cb(Either.unit))
        }
        new StreamSubscription(
          stream,
          subscriber,
```